freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UsersStateDifference.java

```diff
@@ -132,6 +153,42 @@ public static UsersStateDifference forDeletedUser(String deletedUser, Collection
         return usersToAdd;
     }
 
+    public static ImmutableSet<String> calculateUsersToDisable(UmsUsersState umsState, UsersState ipaState) {
+        Map<String, FmsUser> existingIpaUsers = ipaState.getUsers().stream()
+                .collect(Collectors.toMap(FmsUser::getName, Function.identity()));
+        ImmutableSet<String> usersToDisable = ImmutableSet.copyOf(umsState.getUsersState().getUsers().stream()
+                .filter(u -> u.getState() == FmsUser.State.DISABLED &&
+                        existingIpaUsers.containsKey(u.getName()) &&
+                        existingIpaUsers.get(u.getName()).getState() == FmsUser.State.ENABLED)
+                .map(FmsUser::getName)
+                .filter(username -> !FreeIpaChecks.IPA_PROTECTED_USERS.contains(username))
+                .collect(Collectors.toSet()));
+
+        LOGGER.info("usersToDisable size = {}", usersToDisable.size());
+        LOGGER.debug("usersToDisable = {}", usersToDisable);
+
+        return usersToDisable;
+    }
+
+    public static ImmutableSet<String> calculateUsersToEnable(UmsUsersState umsState, UsersState ipaState) {
+        Set<String> enabledUmsUsers = umsState.getUsersState().getUsers().stream()
+                .filter(u -> u.getState() == FmsUser.State.ENABLED)
+                .map(FmsUser::getName)
+                .filter(username -> !FreeIpaChecks.IPA_PROTECTED_USERS.contains(username))
+                .collect(Collectors.toSet());
+        Set<String> disabledIpaUsers = ipaState.getUsers().stream()
+                .filter(u -> u.getState() == FmsUser.State.DISABLED)
+                .map(FmsUser::getName)
+                .collect(Collectors.toSet());
+        ImmutableSet<String> usersToEnable = ImmutableSet
+                .copyOf(Sets.intersection(enabledUmsUsers, disabledIpaUsers));
+
+        LOGGER.info("usersToEnable size = {}", usersToEnable.size());
+        LOGGER.debug("userToEnable = {}", usersToEnable);
+
+        return usersToEnable;
+    }
```

Comment:
these 2 method should use the same logic, and the only difference is whether we are looking for disabled users in freeipa which should be enabled, or enabled users which should be disabled. I think this logic could be in a separate method with parameters. What do you think?

Reply:
thanks! they used to need slightly different logic before I update user-add. I'll take a look and factor it out.

Reply:
updated!